src/services/bunker.py

```diff
@@ -1,15 +1,413 @@
+                if min_bounded_epoch <= e <= epoch:
+                    for v in e_vals:
+                        if v.validator.pubkey not in bounded_slashed_validators:
+                            bounded_slashed_validators[v.index] = int(v.validator.effective_balance)
+            slashings = sum(bounded_slashed_validators.values())
+            adjusted_total_slashing_balance = min(
+                slashings * PROPORTIONAL_SLASHING_MULTIPLIER_BELLATRIX,
+                total_balance
+            )
+            for v in lido_validators_slashed_in_epoch:
+                effective_balance = int(v.validator.effective_balance)
+                penalty_numerator = effective_balance // EFFECTIVE_BALANCE_INCREMENT * adjusted_total_slashing_balance
+                penalty = penalty_numerator // total_balance * EFFECTIVE_BALANCE_INCREMENT
+                per_epoch_lido_midterm_penalties[epoch][v.index] = penalty
+        return per_epoch_lido_midterm_penalties
 
-    def _check_slashings(self):
-        pass
+    @staticmethod
+    def _get_per_frame_lido_midterm_penalties(
+        per_epoch_lido_midterm_penalties: dict[EpochNumber, dict[str, Gwei]],
+        frame_config: FrameConfig,
+    ) -> list[Gwei]:
+        """
+        Put per epoch buckets into per frame buckets to calculate lido midterm penalties impact in each frame
+        """
+        min_lido_midterm_penalty_epoch = min(per_epoch_lido_midterm_penalties.keys())
+        per_frame_buckets: dict[int, dict[int, Gwei]] = defaultdict(dict)
+        left_border = min_lido_midterm_penalty_epoch
+        right_border = frame_config.initial_epoch + frame_config.epochs_per_frame
+        # The first bucket should have flexible size
+        while right_border <= min_lido_midterm_penalty_epoch:
+            right_border += frame_config.epochs_per_frame
+        index = 0
+        for epoch, validator_penalty in sorted(per_epoch_lido_midterm_penalties.items()):
+            if epoch > right_border:
+                index += 1
+                left_border = right_border + 1
+                right_border += frame_config.epochs_per_frame
+            if left_border <= epoch <= right_border:
+                for val_index, penalty in validator_penalty.items():
+                    if val_index not in per_frame_buckets[index]:
+                        per_frame_buckets[index][val_index] = penalty
+        return [sum(penalties.values()) for penalties in per_frame_buckets.values()]
 
-    def _check_activity_leak(self):
-        pass
+    @staticmethod
+    def _calculate_normal_cl_rebase(
+        epochs_passed: int, mean_total_lido_effective_balance: int, mean_total_effective_balance: int
+    ) -> Gwei:
+        normal_cl_rebase = int(
+            NORMALIZED_CL_PER_EPOCH * mean_total_lido_effective_balance * epochs_passed
+            / math.sqrt(mean_total_effective_balance) * (1 - MISTAKE_RANGE)
+        )
+        return Gwei(normal_cl_rebase)
```

Comment:
I didn't figure out the formulas, would appreciate it if you could tell me on the call